### ISSUE TITLE
Need help with custom object training 

### ISSUE THREAD

**Issue by LTU-Eimantas** [NVlabs/Deep_Object_Pose#186]
Hello, I really like the work you have done! But I need help with my model. This is my model.
![1](https://user-images.githubusercontent.com/87859324/134151467-7c2033e2-602a-4b6d-875a-b29e002accee.png)
![2](https://user-images.githubusercontent.com/87859324/134151476-83e3bde2-943b-4a38-b3cf-9483d1078c4a.png)
![3](https://user-images.githubusercontent.com/87859324/134151491-f4802f2c-4dac-4d2c-98cc-c74cc47c974b.png)
![4](https://user-images.githubusercontent.com/87859324/134151508-90827d7a-0152-4d0f-8b39-f5124c847b36.png)

So I trained my model with 90k images and 10k for testing which is made with NVISII and run it 60 epochs. It was painfully slow took about 7 days. I got these tensor images from the training algorithm. 
![0](https://user-images.githubusercontent.com/87859324/134151916-456fe5a8-07fb-49ba-97f6-2c21707f0f8d.png)
![1](https://user-images.githubusercontent.com/87859324/134151926-1a1e59b5-140a-4352-8c8b-e9465a8b704a.png)
![2](https://user-images.githubusercontent.com/87859324/134151933-c0ac3bd2-f142-4b9f-8e3a-11a05e695b53.png)
![3](https://user-images.githubusercontent.com/87859324/134151945-28957018-c4ff-46cf-a0bb-d40cefe13737.png)
![4](https://user-images.githubusercontent.com/87859324/134151956-076c38ed-21ba-4211-bc83-6ba828bddb2d.png)
![5](https://user-images.githubusercontent.com/87859324/134151965-9ee311f4-9c3c-406f-9962-e9d2b3f226f8.png)
![6](https://user-images.githubusercontent.com/87859324/134151978-f6e2381d-971b-41dd-bb45-81bf78ac3a50.png)
![7](https://user-images.githubusercontent.com/87859324/134151991-72aa65a4-9324-4058-9ca8-f5f5e59aa08d.png)
![8](https://user-images.githubusercontent.com/87859324/134151998-2a124755-6e8d-4b4a-b853-8763147ca37b.png)
and  I have run the save option and got these annotations. 
![train_00000](https://user-images.githubusercontent.com/87859324/134152373-7a2deb2c-e2c5-4763-9b90-70c975f9e45b.png)
So after 60 epochs, I got about 0.01-0.009 loss. After running the interface, I got these tensor results 
![beliefMaps2](https://user-images.githubusercontent.com/87859324/134153591-cf25abd9-d583-421c-b026-f692709f8b19.jpg)
![Screenshot from 2021-09-21 13-13-04](https://user-images.githubusercontent.com/87859324/134153611-39e99ed5-f189-4f8f-9d4a-0931af16f135.png)

But if I try to run on the real object I got nothing and this tensor map.
![beliefMaps](https://user-images.githubusercontent.com/87859324/134154368-c44b070c-4196-4615-8615-7e6d592e2d8e.jpg)

So do you have suggestions on how to improve it? It seems like some tensors are good but others don't make sense. One part says it should fix if I feed more data and train more, but it's real times consuming on my rtx 3060 12gb so I thought I might get a second opinion.

**Comment by TontonTremblay** [comment]
These results are highly encouraging, the centroid is well detected. In your training data, do you think the camera field of view is similar to what you are using. Did you try to move away from the robot? It looks like your training data has the robot somewhat far. Also you should train with only a single instance. And since the robot is always going to be on the ground, you could make the poses not as random. I would say 1 or 2 more iteration(s) of data generation. For dope I probably had 10-15 iterations of data generation.  

I am sorry for the 3060 not delivering. I think epoch 20 or 30 should be enough to give you a good idea if it is working.

**Comment by TontonTremblay** [comment]
can you share your 3d model, did you try with nvisii or used ndds?

**Comment by TontonTremblay** [comment]
looking at it again, it seems like there is a symmetry on your model, e.g., the left and right side look similar, is that correct?

**Comment by LTU-Eimantas** [comment]
> can you share your 3d model, did you try with nvisii or used ndds?

It's not full because my engineer team doesn't need all colored models so I added using nvisii and also some materials. 
https://drive.google.com/drive/folders/1p_PYmCjgyNaI-slY8rUb0s4JVfD4nJeH?usp=sharing

**Comment by LTU-Eimantas** [comment]
> looking at it again, it seems like there is a symmetry on your model, e.g., the left and right sides look similar, is that correct?

yeah, they really similar just one button has a different color. But I manage to get some results from these trained wages, but from some perspectives it's isn't working maybe as you said I need to strict rotation axis and make more positions in depth for my training data.
![Screenshot from 2021-09-22 09-22-46](https://user-images.githubusercontent.com/87859324/134394214-9d5c164a-05b1-4bfc-934a-2f0e96a9afa3.png)

**Comment by TontonTremblay** [comment]
Yeah I would just limit the view of the object to one of it side, and do not let it go upside down. Also you can randomly color the robot. We did that in the robot pose estimation work and it helped. Make also the robot appear closer to camera. The testing image you share looks pretty good to me.

**Comment by mintar** [comment]
Here's a script that will modify your existing dataset so that it only shows the object from one side:

https://github.com/NVlabs/Deep_Object_Pose/issues/176#issuecomment-877058216

**Comment by TontonTremblay** [comment]
There is a thing on top of your robot (sorry I just saw this) You should model the thing on top of your robot. But the results are quite good. Good work :P

**Comment by LTU-Eimantas** [comment]
> There is a thing on top of your robot (sorry I just saw this) You should model the thing on top of your robot. But the results are quite good. Good work :P

Yeah, I know it, but I don't need to detect it, only that base and its center because robots going to have different modules on top of it, and training that every time would be really times consuming and not practical.

**Comment by TontonTremblay** [comment]
I am not saying add it to the pose estimation, but if you had it in the training data, the results would be more stable. But overall are you happy with the results?

**Comment by LTU-Eimantas** [comment]
I think it's quite good ( of course could be better :D ), but as a network that is trained basically on synthetic data, I am quite impressed. Of course, further training will be done if the cost of implementation won't be too big, because even on my laptop with gtx 1050 it's struggling to run. Probably Nvidia jetson NX would be required.  But everything depends on the people above me :D .

**Comment by sejmoonwei** [comment]
Hello @LTU-Eimantas .
![image](https://user-images.githubusercontent.com/60536424/144200765-818c8dc7-bf07-428d-84ba-25246d069c8c.png)
Wow , your dataset looks pretty well . How do you deploy the jpg/png file as the background in vnisii? 
Could it be set by dome_light_texture manner ?I found it would cause a spherical deformation on the image on my end.

**Comment by TontonTremblay** [comment]
I have been using dome light. https://github.com/owl-project/NVISII/blob/master/examples/17.materials_visii_interactive.py#L13-L14 

I have downloaded a pretty large set from https://polyhaven.com/hdris This is what I used in the https://arxiv.org/abs/2105.13962 I hope this helps. 

Otherwhise you can use the segmentation mask to do copy and paste onto normal images.

**Comment by sejmoonwei** [comment]
> I have been using dome light. https://github.com/owl-project/NVISII/blob/master/examples/17.materials_visii_interactive.py#L13-L14
> 
> I have downloaded a pretty large set from https://polyhaven.com/hdris This is what I used in the https://arxiv.org/abs/2105.13962 I hope this helps.
> 
> Otherwhise you can use the segmentation mask to do copy and paste onto normal images.

Thank you for helping me again . I'll check this method.

**Comment by TontonTremblay** [comment]
Once I am back from vacation I will try to put a script together to share
here.

**Comment by sejmoonwei** [comment]
Oh I would appreciate it if you can share a template script . Sorry to have disturbed the vacation , do enjoy it .🤣

**Comment by TontonTremblay** [comment]
https://github.com/NVlabs/Deep_Object_Pose/tree/master/scripts/nvisii_data_gen Here is the script I promised. It is pretty bared. Feel free to send PRs.

**Comment by sejmoonwei** [comment]
> https://github.com/NVlabs/Deep_Object_Pose/tree/master/scripts/nvisii_data_gen Here is the script I promised. It is pretty bared. Feel free to send PRs.
Thanks for sharing this work. Though I haven't had a chance to try this script  as I was looking into the loss term , I'll post there if any progress was made.
